File: backend/src/services/mail.service.js
const nodemailer = require('nodemailer');
const { Resend } = require('resend');

const smtpTransporter = nodemailer.createTransport({
  host: process.env.SMTP_HOST,
  port: Number(process.env.SMTP_PORT || 587),
  secure: String(process.env.SMTP_SECURE || 'false').toLowerCase() === 'true',
  auth: process.env.SMTP_USER
    ? {
        user: process.env.SMTP_USER,
        pass: process.env.SMTP_PASS,
      }
    : undefined,
});

let resendClient;

const getResendClient = () => {
  if (!process.env.RESEND_API_KEY) {
    const error = new Error('RESEND_API_KEY is missing from environment variables.');
    error.status = 500;
    throw error;
  }

  if (!resendClient) {
    resendClient = new Resend(process.env.RESEND_API_KEY);
  }

  return resendClient;
};

const isPlainObject = (value) =>
  Boolean(value) && typeof value === 'object' && (value.constructor === Object || Object.getPrototypeOf(value) === null);

const normalizeResendSendResult = (result) => {
  // Resend SDK typically returns: { data, error }
  if (isPlainObject(result) && ('data' in result || 'error' in result)) {
    if (result.error) {
      throwResendError(result.error, 'Resend returned an error.');
    }
    return result.data;
  }

  // Fallback for any older/alternate SDK shapes
  return result;
};

const escapeHtml = (value = '') =>
  String(value)
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#39;');

const sanitizePlainText = (value = '') => String(value).replace(/[\u0000-\u001F\u007F]/g, '').trim();

const throwResendError = (error, fallbackMessage) => {
  const status = error?.statusCode || error?.status || 500;
  const message =
    error?.message ||
    error?.error?.message ||
    fallbackMessage ||
    'Email service failed. Please try again.';

  const wrapped = new Error(message);
  wrapped.status = status;
  wrapped.code = error?.code || error?.name;
  wrapped.details = error?.error || error?.response || error;
  throw wrapped;
};

const sendOtpEmail = async (email, otp) => {
  const provider = String(process.env.AUTH_EMAIL_PROVIDER || 'resend').toLowerCase();
  if (provider === 'smtp') {
    return sendOtpEmailViaSmtp(email, otp);
  }
  return sendOtpEmailViaResend(email, otp);
};

const sendOtpEmailViaResend = async (email, otp) => {
  const from = process.env.RESEND_FROM || 'SecureSend <noreply@securesend.co.in>';

  if (!process.env.RESEND_API_KEY) {
    const error = new Error('RESEND_API_KEY is missing for authentication emails.');
    error.status = 500;
    throw error;
  }

  try {
    const result = await getResendClient().emails.send({
      from,
      to: sanitizePlainText(email),
      subject: 'Your SecureSend Verification Code',
      text: `Your SecureSend verification code is ${otp}. This code will expire in 10 minutes.`,
      html: getOtpEmailHtml(otp),
    });

    const data = normalizeResendSendResult(result);
    if (!data?.id) {
      const error = new Error('Resend did not return a message id. Email may not have been accepted.');
      error.status = 502;
      error.details = data;
      throw error;
    }

    return { provider: 'resend', data };
  } catch (error) {
    throwResendError(error, 'Failed to send verification email via Resend.');
  }
};

const sendOtpEmailViaSmtp = async (email, otp) => {
  try {
    if (!process.env.SMTP_HOST || !process.env.SMTP_USER || !process.env.SMTP_PASS) {
      const error = new Error('SMTP configuration is missing for authentication emails.');
      error.status = 500;
      throw error;
    }

    return await smtpTransporter.sendMail({
      from: `"SecureSend" <${process.env.SMTP_USER}>`,
      to: email,
      subject: 'Your SecureSend Verification Code',
      text: `Your SecureSend verification code is ${otp}. This code will expire in 10 minutes.`,
      html: getOtpEmailHtml(otp),
    });
  } catch (error) {
    throwResendError(error, 'Failed to send verification email.');
  }
};

const getOtpEmailHtml = (otp) => `
  <div style="font-family: sans-serif; max-width: 500px; margin: auto; padding: 20px; border: 1px solid #eee; border-radius: 10px;">
    <h2 style="color: #6366f1; text-align: center;">SecureSend</h2>
    <p>Hello,</p>
    <p>Your verification code is:</p>
    <div style="background: #f4f4f9; padding: 15px; text-align: center; font-size: 24px; font-weight: bold; letter-spacing: 5px; color: #333; border-radius: 8px;">
      ${escapeHtml(otp)}
    </div>
    <p style="font-size: 14px; color: #666; margin-top: 20px;">
      This code will expire in 10 minutes. If you didn't request this, please ignore this email.
    </p>
  </div>
`;

const sendAnonymousEmail = async ({ to, subject, content, alias }) => {
  const safeTo = sanitizePlainText(to);
  const safeSubject = sanitizePlainText(subject);
  const safeContent = sanitizePlainText(content);
  const safeAlias = sanitizePlainText(alias);

  // Validate required fields
  if (!safeTo || !safeSubject || !safeContent) {
    const error = new Error('Missing required email fields.');
    error.status = 400;
    throw error;
  }

  if (!safeAlias) {
    const error = new Error('Alias is required for anonymous emails.');
    error.status = 400;
    throw error;
  }

  if (!/^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(safeTo)) {
    const error = new Error('Please enter a valid recipient email address.');
    error.status = 400;
    throw error;
  }

  console.log('[sendAnonymousEmail] Processing anonymous email', {
    alias: safeAlias,
    to: safeTo,
    subject: safeSubject,
  });

  // Choose email provider
  const provider = String(process.env.ANON_EMAIL_PROVIDER || '').toLowerCase();
  if (provider === 'smtp') {
    return sendAnonymousEmailViaSmtp({ to: safeTo, subject: safeSubject, content: safeContent, alias: safeAlias });
  }

  if (!process.env.RESEND_API_KEY) {
    // If Resend isn't configured, fall back to SMTP when available.
    return sendAnonymousEmailViaSmtp({ to: safeTo, subject: safeSubject, content: safeContent, alias: safeAlias });
  }

  // Use Resend provider
  return sendAnonymousEmailViaResend({
    to: safeTo,
    subject: safeSubject,
    content: safeContent,
    alias: safeAlias,
  });
};

/**
 * Send anonymous email via Resend
 * From: SecureSend (configured via RESEND_FROM in .env)
 * ReplyTo: alias@securesend.co.in (alias-based reply address)
 */
const sendAnonymousEmailViaResend = async ({ to, subject, content, alias }) => {
  const safeHtml = escapeHtml(content).replace(/\n/g, '<br />');

  // Use RESEND_FROM from environment (should be verified domain)
  const from = process.env.RESEND_FROM || 'noreply@securesend.co.in';

  if (!from) {
    const error = new Error(
      'RESEND_FROM is not set. Configure a verified sender in Resend and set RESEND_FROM in .env'
    );
    error.status = 500;
    throw error;
  }

  // Format reply-to as alias@securesend.co.in
  const replyTo = `${alias}@securesend.co.in`;

  try {
    console.log('[sendAnonymousEmailViaResend] Preparing to send', {
      from,
      to,
      subject,
      replyTo,
      contentLength: content.length,
      timestamp: new Date().toISOString(),
    });

    const result = await getResendClient().emails.send({
      from,
      to,
      subject: `Anonymous: ${subject || 'New Message'}`,
      replyTo,
      text: content,
      html: `
        <div style="font-family:sans-serif;padding:20px;color:#333">
          <h2>📩 You received an anonymous message</h2>
          <div style="background:#f5f5f5;padding:15px;border-radius:8px;margin:15px 0">
            ${safeHtml}
          </div>
          <hr style="border:none;border-top:1px solid #ddd;margin:20px 0"/>
          <small style="color:#666">
            Sent via <strong>SecureSend</strong> • Reply to: <code>${replyTo}</code>
          </small>
        </div>
      `,
    });

    const data = normalizeResendSendResult(result);
    const messageId = data?.id;

    console.log('[sendAnonymousEmailViaResend] Response received', {
      messageId,
      alias,
      to,
      status: data?.status,
      createdAt: data?.created_at,
    });

    if (!messageId) {
      const error = new Error('Resend did not return a message id. Email may not have been accepted.');
      error.status = 502;
      error.details = data;
      console.error('[sendAnonymousEmailViaResend] No message ID in response:', {
        data,
        to,
      });
      throw error;
    }

    console.log('[sendAnonymousEmailViaResend] Email sent successfully', {
      messageId,
      alias,
      to,
    });

    return { provider: 'resend', data };
  } catch (error) {
    console.error('[sendAnonymousEmailViaResend] Error sending email:', {
      error: error.message,
      alias,
      to,
      status: error.status,
    });
    throwResendError(error, 'Failed to send anonymous email via Resend. Please try again.');
  }
};

/**
 * Send anonymous email via SMTP
 * From: Configured SMTP sender
 * ReplyTo: alias@securesend.co.in (alias-based reply address)
 */
const sendAnonymousEmailViaSmtp = async ({ to, subject, content, alias }) => {
  try {
    if (!process.env.SMTP_HOST || !process.env.SMTP_USER || !process.env.SMTP_PASS) {
      const error = new Error('SMTP configuration is missing for sending emails.');
      error.status = 500;
      throw error;
    }

    const safeTo = sanitizePlainText(to);
    const safeSubject = sanitizePlainText(subject);
    const safeContent = sanitizePlainText(content);
    const safeAlias = sanitizePlainText(alias);

    // Format reply-to as alias@securesend.co.in
    const replyTo = `${safeAlias}@securesend.co.in`;

    console.log('[sendAnonymousEmailViaSmtp] Preparing to send via SMTP', {
      from: process.env.SMTP_USER,
      to: safeTo,
      subject: safeSubject,
      replyTo,
      timestamp: new Date().toISOString(),
    });

    const result = await smtpTransporter.sendMail({
      from: `"SecureSend" <${process.env.SMTP_USER}>`,
      to: safeTo,
      subject: `Anonymous: ${safeSubject || 'New Message'}`,
      replyTo,
      text: safeContent,
      html: `
        <div style="font-family:sans-serif;padding:20px;background:#f9f9f9;border-radius:8px">
          <h2 style="color:#333">📩 You received an anonymous message</h2>
          <div style="background:white;padding:15px;border-radius:4px;margin:20px 0">
            <p style="margin:0;white-space:pre-wrap;color:#555">${escapeHtml(safeContent).replace(/\n/g, '<br>')}</p>
          </div>
          <hr style="border:none;border-top:1px solid #ddd;margin:20px 0"/>
          <small style="color:#999">Sent via <strong>SecureSend</strong> • Reply to: <code>${replyTo}</code></small>
        </div>
      `,
    });

    console.log('[sendAnonymousEmailViaSmtp] Email sent successfully', {
      messageId: result.messageId,
      to: safeTo,
      alias: safeAlias,
      response: result.response,
    });

    return { provider: 'smtp', data: result };
  } catch (error) {
    console.error('[sendAnonymousEmailViaSmtp] Error sending email:', {
      error: error.message,
      code: error.code,
      to,
      alias,
    });
    throwResendError(error, 'Failed to send anonymous email via SMTP. Please try again.');
  }
};

module.exports = { sendOtpEmail, sendAnonymousEmail };

File: backend-java/src/main/java/com/securesend/service/MailService.java
package com.securesend.service;

import com.securesend.dto.AnonymousRequests.AttachmentDto;
import com.securesend.exception.ApiException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.*;
import org.springframework.mail.javamail.JavaMailSender;
import org.springframework.mail.javamail.MimeMessageHelper;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import jakarta.mail.internet.MimeMessage;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@Service
public class MailService {

    private static final Logger log = LoggerFactory.getLogger(MailService.class);

    @Value("${securesend.resend.api-key:}")
    private String resendApiKey;

    @Value("${securesend.resend.from:SecureSend <noreply@securesend.co.in>}")
    private String resendFrom;

    @Value("${securesend.email.auth-provider:resend}")
    private String authProvider;

    @Value("${securesend.email.anon-provider:resend}")
    private String anonProvider;

    private final JavaMailSender javaMailSender;
    private final RestTemplate restTemplate;

    public MailService(JavaMailSender javaMailSender) {
        this.javaMailSender = javaMailSender;
        this.restTemplate = new RestTemplate();
    }

    public void sendOtpEmail(String email, String otp) {
        if ("smtp".equalsIgnoreCase(authProvider)) {
            sendOtpEmailViaSmtp(email, otp);
        } else {
            sendOtpEmailViaResend(email, otp);
        }
    }

    private void sendOtpEmailViaResend(String email, String otp) {
        if (resendApiKey == null || resendApiKey.isBlank()) {
            throw new ApiException("RESEND_API_KEY is missing from environment variables.", HttpStatus.INTERNAL_SERVER_ERROR);
        }

        String htmlContent = String.format("""
            <div style="font-family: sans-serif; max-width: 500px; margin: auto; padding: 20px; border: 1px solid #eee; border-radius: 10px;">
              <h2 style="color: #6366f1; text-align: center;">SecureSend</h2>
              <p>Hello,</p>
              <p>Your verification code is:</p>
              <div style="background: #f4f4f9; padding: 15px; text-align: center; font-size: 24px; font-weight: bold; letter-spacing: 5px; color: #333; border-radius: 8px;">
                %s
              </div>
              <p style="font-size: 14px; color: #666; margin-top: 20px;">
                This code will expire in 10 minutes. If you didn't request this, please ignore this email.
              </p>
            </div>
            """, escapeHtml(otp));

        Map<String, Object> body = new HashMap<>();
        body.put("from", resendFrom);
        body.put("to", List.of(email.trim()));
        body.put("subject", "Your SecureSend Verification Code");
        body.put("text", "Your SecureSend verification code is " + otp + ". This code will expire in 10 minutes.");
        body.put("html", htmlContent);

        sendViaResendHttp(body);
    }

    private void sendOtpEmailViaSmtp(String email, String otp) {
        try {
            MimeMessage mimeMessage = javaMailSender.createMimeMessage();
            MimeMessageHelper helper = new MimeMessageHelper(mimeMessage, true, "UTF-8");
            helper.setFrom(resendFrom);
            helper.setTo(email);
            helper.setSubject("Your SecureSend Verification Code");
            helper.setText("Your SecureSend verification code is " + otp + ". This code will expire in 10 minutes.", true);
            javaMailSender.send(mimeMessage);
        } catch (Exception e) {
            log.error("Failed to send OTP via SMTP", e);
            throw new ApiException("Failed to send verification email via SMTP.", HttpStatus.INTERNAL_SERVER_ERROR);
        }
    }

    public Map<String, Object> sendAnonymousEmail(String to, String subject, String content, String alias, List<AttachmentDto> attachments) {
        String safeTo = to.trim();
        String safeSubject = subject.trim();
        String safeContent = content.trim();
        String safeAlias = alias.trim();

        String customDisplayName = "SecureSend";
        String cleanContent = safeContent;

        Pattern fromPattern = Pattern.compile("^\\[From:\\s*([^\\]]+)\\]", Pattern.CASE_INSENSITIVE);
        Matcher matcher = fromPattern.matcher(safeContent);
        if (matcher.find()) {
            customDisplayName = matcher.group(1).trim();
            cleanContent = safeContent.replaceAll("^\\[From:\\s*[^\\]]+\\]\\s*", "");
        }

        if ("smtp".equalsIgnoreCase(anonProvider) || resendApiKey == null || resendApiKey.isBlank()) {
            return sendAnonymousEmailViaSmtp(safeTo, safeSubject, cleanContent, safeAlias, customDisplayName, attachments);
        }
        return sendAnonymousEmailViaResend(safeTo, safeSubject, cleanContent, safeAlias, customDisplayName, attachments);
    }

    private Map<String, Object> sendAnonymousEmailViaResend(String to, String subject, String content, String alias, String displayName, List<AttachmentDto> attachments) {
        String replyTo = alias + "@securesend.co.in";
        String emailOnly = resendFrom;
        if (resendFrom.contains("<") && resendFrom.contains(">")) {
            emailOnly = resendFrom.substring(resendFrom.indexOf("<") + 1, resendFrom.indexOf(">"));
        }
        String fromHeader = String.format("\"%s\" <%s>", displayName, emailOnly);

        String safeHtml = escapeHtml(content).replace("\n", "<br />");
        String htmlContent = String.format("""
            <div style="font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif; font-size: 14px; line-height: 1.6; color: #222; max-width: 600px;">
              <div style="margin-bottom: 24px;">
                %s
              </div>
              <hr style="border: none; border-top: 1px solid #eee; margin: 24px 0 16px 0;" />
              <div style="font-size: 11px; color: #888;">
                Sent securely via <strong>SecureSend</strong>. You can reply directly to: <code>%s</code>
              </div>
            </div>
            """, safeHtml, replyTo);

        Map<String, Object> body = new HashMap<>();
        body.put("from", fromHeader);
        body.put("to", List.of(to));
        body.put("subject", subject != null && !subject.isBlank() ? subject : "New Message");
        body.put("reply_to", replyTo);
        body.put("text", content);
        body.put("html", htmlContent);

        if (attachments != null && !attachments.isEmpty()) {
            List<Map<String, String>> resendAtts = new ArrayList<>();
            for (AttachmentDto att : attachments) {
                Map<String, String> a = new HashMap<>();
                a.put("filename", att.getFilename());
                a.put("content", att.getContent()); // Base64 content
                resendAtts.add(a);
            }
            body.put("attachments", resendAtts);
        }

        Map<String, Object> responseData = sendViaResendHttp(body);
        Map<String, Object> result = new HashMap<>();
        result.put("provider", "resend");
        result.put("data", responseData);
        return result;
    }

    private Map<String, Object> sendAnonymousEmailViaSmtp(String to, String subject, String content, String alias, String displayName, List<AttachmentDto> attachments) {
        try {
            String replyTo = alias + "@securesend.co.in";
            MimeMessage mimeMessage = javaMailSender.createMimeMessage();
            MimeMessageHelper helper = new MimeMessageHelper(mimeMessage, true, "UTF-8");
            helper.setFrom(resendFrom, displayName);
            helper.setReplyTo(replyTo);
            helper.setTo(to);
            helper.setSubject(subject != null && !subject.isBlank() ? subject : "New Message");
            helper.setText(content, true);

            javaMailSender.send(mimeMessage);

            Map<String, Object> result = new HashMap<>();
            result.put("provider", "smtp");
            result.put("data", Map.of("status", "sent"));
            return result;
        } catch (Exception e) {
            log.error("SMTP anonymous mail send failed", e);
            throw new ApiException("Failed to send anonymous email via SMTP.", HttpStatus.INTERNAL_SERVER_ERROR);
        }
    }

    private Map<String, Object> sendViaResendHttp(Map<String, Object> payload) {
        try {
            HttpHeaders headers = new HttpHeaders();
            headers.setContentType(MediaType.APPLICATION_JSON);
            headers.setBearerAuth(resendApiKey);

            HttpEntity<Map<String, Object>> entity = new HttpEntity<>(payload, headers);
            ResponseEntity<Map> response = restTemplate.exchange("https://api.resend.com/emails", HttpMethod.POST, entity, Map.class);

            if (response.getStatusCode().is2xxSuccessful() && response.getBody() != null) {
                return (Map<String, Object>) response.getBody();
            }
            throw new ApiException("Resend returned status: " + response.getStatusCode(), HttpStatus.BAD_GATEWAY);
        } catch (Exception e) {
            log.error("Resend API request failed", e);
            throw new ApiException("Failed to send email via Resend: " + e.getMessage(), HttpStatus.BAD_GATEWAY);
        }
    }

    private String escapeHtml(String text) {
        if (text == null) return "";
        return text.replace("&", "&amp;")
                .replace("<", "&lt;")
                .replace(">", "&gt;")
                .replace("\"", "&quot;")
                .replace("'", "&#39;");
    }
}
